src/spikeinterface/preprocessing/common_reference.py

```diff
@@ -180,34 +180,60 @@ def __init__(
-
-        if self.reference == "global":
-            for chan_inds, chan_group_inds in self._groups(_channel_indices):
-                out_inds = np.array([np.where(_channel_indices == i)[0][0] for i in chan_inds])
-                out_traces[:, out_inds] = all_traces[:, chan_inds] - self.operator_func(all_traces[:, chan_group_inds])
-
-        elif self.reference == "single":
-            for i, (chan_inds, _) in enumerate(self._groups(_channel_indices)):
-                out_inds = np.array([np.where(_channel_indices == i)[0][0] for i in chan_inds])
-                out_traces[:, out_inds] = all_traces[:, chan_inds] - self.operator_func(
-                    all_traces[:, [self.ref_channel_indices[i]]]
-                )
-
-        elif self.reference == "local":
-            for i, chan_ind in enumerate(_channel_indices):
-                out_traces[:, [i]] = all_traces[:, [chan_ind]] - self.operator_func(
-                    all_traces[:, self.neighbors[chan_ind]]
-                )
-        return out_traces
+        # Let's do the case with group_indices equal None as that is easy
+        if self.group_indices is None:
+            operator = np.mean if self.operator == "average" else np.median
+
+            # We need all the channels to calculate the reference
+            traces = self.parent_recording_segment.get_traces(start_frame, end_frame, slice(None))
+
+            if self.reference == "global":
+                shift = operator(traces, axis=1, keepdims=True)
+                re_referenced_traces = traces[:, channel_indices] - shift[:, channel_indices]
+            elif self.reference == "single":
+                shift = operator(traces[:, self.ref_channel_indices], axis=1, keepdims=True)
+                re_referenced_traces = traces[:, channel_indices] - shift[:, channel_indices]
+            elif self.reference == "local":
+                re_referenced_traces = np.zeros_like(traces[:, channel_indices])
+                for channel_index, channel_neighbords in self.neighbors.items():
+                    channel_shift = operator(traces[:, channel_neighbords], axis=1)
+                    re_referenced_traces[:, channel_index] = traces[:, channel_index] - channel_shift
+
+            return re_referenced_traces.astype(self.dtype, copy=False)
+
+        # Then the old implementation for backwards compatibility that supports grouping
+        else:
+            # need input trace
+            all_traces = self.parent_recording_segment.get_traces(start_frame, end_frame, slice(None))
+            all_traces = all_traces.astype(self.dtype)
+
+            self.temp = np.zeros((all_traces.shape[0],), dtype=all_traces.dtype)
+            _channel_indices = np.arange(all_traces.shape[1])
+            if channel_indices is not None:
+                _channel_indices = _channel_indices[channel_indices]
+
+            out_traces = np.zeros((all_traces.shape[0], _channel_indices.size), dtype=self.dtype)
```

Comment:
@h-mayorquin 

After some discussion, probably the fix could be very simple and without the if-else `group_indices` with 2 changes:
1. In the `self.operator_func` remove `out=self.temp` and add `keepdims=True`
2. Do not pre allocate `out_traces`
3. Add astype at the end in any case.

I'd be happy to implement this in a separate PR!

Reply:
Up to you guys,

I do believe there is a benefit on separting the non-grouping code to the grouping code.  The former can be made way simpler and will be easier for a future developer to understand, modify and optimize, the later has to be more complicated and will be more difficult. 

I recommend that you add a regression test whatever you decide to do though.

Reply:
sounds good, then I'll push the same logic to the grouping branch, so it's fixed there too!